### PR TITLE
Modified extension gadget

### PR DESCRIPTION
**Pull request by miha-stopar**
### Description

This PR will replace old modified extension node PRs (there were multiple because the witness generator was in a separate repository at that time).

Support for modified extension nodes.

This happens when an extension node is replaced by another (shorter or longer) extension node. For example, we have an extension node at `n1 n2 n3 n4 n5 n6` with branch with two leaves at positions `n` and `m`. If we add a leaf at `n1 n2 n3 n4 n7` where `n5 != n7`, a new extension node is inserted at `n1 n2 n3 n4` with a new branch with an extension node at position `n5` (with only one nibble `n6`) and a leaf at position `n7`. In this case, the S proof contains the extension node at `n1 n2 n3 n4 n5 n6` and no underlying branch and leaf (the modification happens at `n1 n2 n3 n4 n7` and only an extension node is find there), thus we need to add a placeholder branch and a placeholder leaf.

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: mpt-witness-generator/witness/leaf.go
Comment (ChihChengLiang) [inline mpt-witness-generator/witness/leaf.go:541]: Can we explain this?

**Comment by miha-stopar** [comment]
> First round of review on the witness generator part.
> I'll review the circuit in the next round.

Thanks @ChihChengLiang ! I think addressed all the comments in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685/commits/67399e9a144929c6f9e90c02bfb7790706bca00a.

File: zkevm-circuits/src/mpt_circuit/extension.rs
Comment (miha-stopar) [reply zkevm-circuits/src/mpt_circuit/extension.rs:74]: An example extension node witness is:
```
[130 0 150 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
[160 215 121 207 40 14 160 149 115 175 222 139 45 208 88 81 65 226 190 111 191 208 252 147 90 105 163 154 4 132 52 204 121 0]
[0 0 6 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
[160 208 13 115 5 131 97 180 161 37 184 30 176 46 254 113 83 244 235 231 202 172 66 172 99 25 151 249 73 32 160 1 218 0]
```

The third row (`ExtensionBranchRowType::Nibbles`) contains the second nibbles of the key (the first row). So we want to have the range check `0 - 15` which is done if the row is marked by `ExtensionBranchRowType::Nibbles`.

File: zkevm-circuits/src/mpt_circuit/extension.rs
Comment (miha-stopar) [reply zkevm-circuits/src/mpt_circuit/extension.rs:162]: I moved the parity update down to the line 176 to have the information about the parity _before_ the extension node in `ext_key_rlc_calc_value`. Please see the comment below.

File: zkevm-circuits/src/mpt_circuit/mod_extension.rs
Comment (miha-stopar) [reply zkevm-circuits/src/mpt_circuit/mod_extension.rs:100]: Yes, totally! Just needed to add `.expr()` after each word, otherwise getting:
```
no method named `mul` found for tuple `(circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>)`
```

File: zkevm-circuits/src/mpt_circuit/mod_extension.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/mpt_circuit/mod_extension.rs:27]: Let me know if the doc makes sense
```suggestion
/// Validity conditions for ModExtensionGadget:
/// 1. Both the long and short extension nodes are in the parent node.
/// 2. Both the branches of the short extension node’s and the long extension node’s must be the same
/// 3. The concatenation of the middle extension node and short ext node must equal the long ext node
```

**Review by ChihChengLiang — COMMENTED** [review]
This is my final round of review.
I'll approve after the feedback is addressed.

File: zkevm-circuits/src/mpt_circuit/account_leaf.rs
Comment (miha-stopar) [reply zkevm-circuits/src/mpt_circuit/account_leaf.rs:283]: As expected the block is executed when the condition is true.

For passing the condition to the gadget - that would mean having one big if statement in `ModExtensionGadget` or having the condition in each constraint statement. I think the current code is somehow easier to read?

File: zkevm-circuits/src/mpt_circuit/helpers.rs
Comment (Brechtpd) [inline zkevm-circuits/src/mpt_circuit/helpers.rs:264]:
```suggestion
            and::expr(&[is_long.expr(), not!(xor::expr(is_key_odd.expr(), is_key_part_odd.expr()))]) => {
```

Also helps to make it more clear it's the `else` clause. 

Hmmm maybe it's even better if it's split up in `is_long` and `is_short` cases, and then the `is_long` case contains an `if xor::expr(is_key_odd.expr(), is_key_part_odd.expr()) { } else { }`.

Comment (miha-stopar) [reply zkevm-circuits/src/mpt_circuit/helpers.rs]: Changed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685/commits/4acc57394e92f71d24b4622c78d885a01a698a3d

**Review by ChihChengLiang — APPROVED** [review]
~~The rest parts LGTM. I still have some concerns on [this part](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685#discussion_r1425598669), but plan to create a minimal example to demonstrate the effect after the merging of this PR.~~

Update: LGTM. Concerns resolved in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685#discussion_r1437140715

File: zkevm-circuits/src/mpt_circuit/account_leaf.rs
Comment (Brechtpd) [reply zkevm-circuits/src/mpt_circuit/account_leaf.rs:283]: Any specific concerns here about the use of `ifx!` @ChihChengLiang ? I think `ifx` is being used here similarly to other places, so not sure how the specific concerns here.

Some of the gadgets are created inside the `ifx!` because otherwise things also get quite a bit messier to keep track of all the different paths and inputs, but regardless of the condition they are always created.

Comment (ChihChengLiang) [reply zkevm-circuits/src/mpt_circuit/account_leaf.rs:283]: Hi @Brechtpd @miha-stopar,
I've managed to convince myself that the current way of `ifx!` works with this mini example https://github.com/privacy-scaling-explorations/zkevm-circuits/commit/1dd12cadca0094884010724af12dffb8f2887ba7. The concern is resolved. Also thank you all for explaining to me.

**Pull request merged**
